lint/rule.go

```diff
@@ -14,10 +14,295 @@
 
 package lint
 
+import (
+	"fmt"
+	"strings"
+
+	"github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/jhump/protoreflect/desc"
+)
+
 // Rule defines a lint rule that checks Google Protobuf APIs.
 type Rule interface {
-	// Info returns metadata about a rule.
-	Info() RuleInfo
-	// Lint performs the linting process.
-	Lint(Request) ([]Problem, error)
+	// GetName returns the name of the rule.
+	GetName() RuleName
+
+	// GetURI returns the URI where the applicable guideline
+	// is documented. (This should generally be an AIP on https://aip.dev/.)
+	GetURI() string
```

Comment:
What is the purpose of this function? If you want to provide information about a rule in command lines, please note that a AIP (or a sub section) would contains multiple rules. For example, from the URI https://aip.dev/131#request-message, I am not certain if a rule is about "A resource name field must be included" or "It should be called name".

Reply:
Internally, we should be including the URI at the end of the message we post in Critique.
If we can not do that, we should just start putting this in the problem messages.

Reply:
I think we should put the URI into the problem messages, as the URI does not serve as an identity, but more like "for more context/details, see this link".

Reply:
Should we give some description about a rule, i.e., `Description() string`, so a user can understand a rule before running and getting the problem messages (which may be tricky to trigger one before understanding the rule)?

Reply:
URI is much more valuable than description (which can just be a comment above the rule). If we should not have URI, we _definitely_ should not have description.

Reply:
Maybe we should start to discuss about having rule numbers in an AIP, e.g., AIP-131-R1 for the rule that checks that "A resource name field must be included", and in the AIP text, a reference to the rule. We may want to remove RuleName then, since sometimes it is really hard to give a unique, informative name.

Reply:
> Maybe we should start to discuss about having rule numbers in an AIP, e.g., AIP-131-R1 for the rule that checks that "A resource name field must be included"

This is what I originally asked for.

That said, at this point, I would like to get this PR in, as it is going to keep getting larger. If we want to refactor rule names, let's do a separate followup.

Reply:
Totally agree! Thanks!